pytest/backend/test_states_polyquad.py

```diff
@@ -0,0 +1,466 @@
+        backend.prepare_displaced_squeezed_state(a, r, phi, 0)
+
+        state = backend.state()
+        mean, var = state.poly_quad_expectation(A, d, k, phi=qphi)
+
+        mean_ex, var_ex = sample_normal_expectations(
+            lambda X, P, XP: XP, correction=-np.linalg.det(2 * A[:, [0, 3]][[0, 3]])
+        )
+        assert np.allclose(mean, mean_ex, atol=tol, rtol=0)
+        assert np.allclose(var, var_ex, atol=tol, rtol=0)
+
+    def test_arbitrary_quadratic(
+        self, setup_backend, tol, pure, sample_normal_expectations
+    ):
+        """Test that the correct result is returned for E(c0 x^2 + c1 p^2 + c2 xp + c3 x + c4 p + k)"""
+        backend = setup_backend(3)
+        backend.reset(cutoff_dim=12, pure=pure)
+
+        c0 = 1 / np.sqrt(2)
+        c1 = 3
+        c2 = 0.53
+        c3 = 1 / 3.0
+        c4 = -1
+
+        # define the arbitrary quadratic
+        A = np.zeros([6, 6])
+        A[0, 0] = c0
+        A[3, 3] = c1
+
+        A[3, 0] = c2 / 2
+        A[0, 3] = c2 / 2
+
+        # define the linear combination and constant term
+        d = np.array([c3, 0, 0, c4, 0, 0])
+        k = 5
+
+        # prepare a squeezed displaced state
+        backend.prepare_displaced_squeezed_state(a, r, phi, 0)
+
+        state = backend.state()
+        mean, var = state.poly_quad_expectation(A, d, k, phi=qphi)
+
+        mean_ex, var_ex = sample_normal_expectations(
+            lambda X, P, XP: c0 * X ** 2 + c1 * P ** 2 + c2 * XP + c3 * X + c4 * P + k,
+            correction=-np.linalg.det(2 * A[:, [0, 3]][[0, 3]]),
+        )
+
+        assert np.allclose(mean, mean_ex, atol=tol, rtol=0)
+        assert np.allclose(var, var_ex, atol=tol, rtol=0)
+
+
+@pytest.mark.backends("fock", "gaussian")
+class TestMultiModePolyQuadratureExpectations:
+    """Test multi mode poly_quad_expectation methods"""
+
+    def test_three_mode_arbitrary(self, setup_backend, pure, hbar, tol):
+        """Test that the correct result is returned for an arbitrary quadratic polynomial"""
+        backend = setup_backend(3)
+        backend.reset(cutoff_dim=7, pure=pure)
+
+        # fmt:off
+        A = np.array([[ 0.7086495 , -0.39299695,  0.30536448,  0.48822049,  0.64987373, 0.7020327 ],
+                      [-0.39299695,  0.0284145 ,  0.53202656,  0.20232385,  0.26288656, 0.20772833],
+                      [ 0.30536448,  0.53202656,  0.28126466,  0.64192545, -0.36583748, 0.51704656],
+                      [ 0.48822049,  0.20232385,  0.64192545,  0.51033017,  0.29129713, 0.77103581],
+                      [ 0.64987373,  0.26288656, -0.36583748,  0.29129713,  0.37646972, 0.2383589 ],
+                      [ 0.7020327 ,  0.20772833,  0.51704656,  0.77103581,  0.2383589 ,-0.96494418]])
+
+        # fmt:on
+        d = np.array(
+            [0.71785224, -0.80064627, 0.08799823, 0.76189805, 0.99665321, -0.60777437]
+        )
+        k = 0.123
+
+        a_list = [0.044 + 0.023j, 0.0432 + 0.123j, -0.12 + 0.04j]
```

Comment:
I've been thinking a bit about this. I was reading somewhere that all test parameters should live inside the test function, so that if copied/pasted it should run and pass

Reply:
You mean parameters that are specific to a single particular test?

Reply:
Yep.